File: ML-Practicas/P2 - Linear Regresion/multiple_linear_regression.R
#regresión lineal multiple

dataset=read.csv('50_Startups.csv')
#dataset=dataset[,2:3]

#Vriable categorica a variables dummy (En R son factores)

dataset$State=factor(dataset$State,
                       levels=c("New York","California","Florida"),
                       labels=c(1,2,3))

#dividir los datos en conjuntos de entrenamiento y test
#install.packages("caTools")
library(caTools)
set.seed(123) #Semilla para aleatorios
split=sample.split(dataset$Profit,SplitRatio =0.8)

training_set=subset(dataset,split==TRUE)
testing_set=subset(dataset,split==FALSE)

# Escalado de valores
# training_set[,2:3]=scale(training_set[,2:3])
# testing_set[,2:3]=scale(testing_set[,2:3])

#Ajustar modelo de regresión Lineal Multiple con el Conjunto de Entrenamiento

# regression=lm(formula=Profit ~ R.D.Spend + Administration + Marketing.Spend + State,data=training_set)
regression=lm(formula=Profit ~ ., data=training_set)

# Predecir los resultados con el Conjunto de Test

y_pred=predict(regression,newdata = testing_set)

#Construir un modelo optimo con la eliminación hacia atrás
SL=0.05 #Nivel de significancia es arbitrario, lo normal es manejar 0.05

regression=lm(formula=Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
              data=dataset)
summary(regression)

#Borrado state por un P valor alto,mayor a 0.05
regression=lm(formula=Profit ~ R.D.Spend + Administration + Marketing.Spend,
              data=dataset)
summary(regression)

#Borrado Administration por un P valor alto,mayor a 0.05
regression=lm(formula=Profit ~ R.D.Spend + Marketing.Spend,
              data=dataset)
summary(regression)

#Borrado Marketing.Spend  por un P valor alto,mayor a 0.05 ( tene un valor de 0.6, que podemos utilizar y pasarlo por alto)
regression=lm(formula=Profit ~ R.D.Spend,
              data=dataset)
summary(regression)

File: ML-Practicas/P2 - Linear Regresion/multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jun 17 13:28:22 2019
@author: adrian.perez
"""
# Regresión Linear Multiple

# Plantilla de Preprocesado 
#Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importar el dataset
dataset = pd.read_csv("50_startups.csv")

X=dataset.iloc[:, :-1].values #EStablece las variables independientes
y=dataset.iloc[:, 4].values #establecer variable dependiente


#Codificar datos categoricos para el País en variables Dummy
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X=LabelEncoder()
X[:,3]=labelencoder_X.fit_transform(X[:,3])
onehotencoder=OneHotEncoder(categorical_features=[3])
X=onehotencoder.fit_transform(X).toarray()

#Evitar la trampa de las variables Dummy
X=X[:,1:]

#Dividir el dataset en conjuto de entrenamiento y conjunto de test
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=.20,random_state=0)

#Ajustar el modelo de regresión lienal multiple con daots deTraining 
from sklearn.linear_model import LinearRegression
regression=LinearRegression()
regression.fit(X_train,y_train)

#Predicción de los resultados con el conjunto de testing

y_pred=regression.predict(X_test)

#construir modelo optimo de regrisiónm multiple utilizando la eliminación hacia atrs
import statsmodels.formula.api as sm
X=np.append(arr=np.ones((50,1)).astype(int),values=X,axis=1)
SL=0.05

X_opt=X[:,[0,1,2,3,4,5]]
regression_OLS=sm.OLS(endog=y,exog=X_opt).fit()
regression_OLS.summary()

X_opt=X[:,[0,1,3,4,5]]
regression_OLS=sm.OLS(endog=y,exog=X_opt).fit()
regression_OLS.summary()

X_opt=X[:,[0,3,4,5]]
regression_OLS=sm.OLS(endog=y,exog=X_opt).fit()
regression_OLS.summary()

X_opt=X[:,[0,3,5]]
regression_OLS=sm.OLS(endog=y,exog=X_opt).fit()
regression_OLS.summary()

X_opt=X[:,[0,3]]
regression_OLS=sm.OLS(endog=y,exog=X_opt).fit()
regression_OLS.summary()



